### PR TITLE
Added support for Arbitrary

### LINKED ISSUE #122: impl Arbitrary for BigDecimal
Hey!

I'm trying to fuzz some code that uses BigDecimal, and am hitting the issue that it does not implement `Arbitrary`. Currently, my solution is the following code to generate one:
```rust
BigDecimal::new(u.arbitrary()?, u.arbitrary::<i64>()? % 1000000)
```

I originally did not have the `% 1000000`, but it resulted in memory allocation failures.

What's the recommended way to use BigDecimal from a fuzzer? Relatedly, maybe it'd make sense to add an `arbitrary` feature to BigDecimal, allowing to auto-generate them?

Anyway, thank you for BigDecimal! I'm currently stuck on 0.3 because sqlx didn't update probably due to https://github.com/akubera/bigdecimal-rs/issues/117 ; but it's really coming in handy to not have to care about floating points and integers too much :)

### PR DESCRIPTION
**Pull request by LucaCappelletti94**
This PR adds support for the [`arbitrary`](https://crates.io/crates/arbitrary) crate by implementing the `Arbitrary` trait for `BigDecimal`. This enables using `BigDecimal` with fuzzing tools like `cargo-fuzz` and `libfuzzer`, and more importantly allows to use the `Arbitrary` derive in third party structs which include this struct as a field.

A new optional feature `fuzz` has been added to `Cargo.toml`:

```toml
[features]
fuzz = ["std", "arbitrary", "num-bigint/arbitrary"]
```

**Note:** The feature is named `fuzz` (not `arbitrary`) to avoid a name collision with the dependency, which is required for compatibility with older Cargo versions (pre-1.60). The feature requires `std` because the `arbitrary` crate itself does not support `no_std`

## Implementation Details

To prevent memory allocation failures during fuzzing, the implementation limits the generated scale to `±1,000,000`. Without this limit, extremely large scale values could cause the `BigDecimal` to attempt allocating massive amounts of memory when converted to strings or performing certain operations.

- **New module:** `src/impl_arbitrary.rs` implementing `Arbitrary<'a>` for `BigDecimal`
- Implements all three trait methods:
  - `arbitrary()` - Generate a random `BigDecimal` from unstructured data
  - `arbitrary_take_rest()` - Consume all remaining data to generate a `BigDecimal`
  - `size_hint()` - Provide size hints for the fuzzer

## Usage

Enable the feature in your `Cargo.toml`:

```toml
[dependencies]
bigdecimal = { version = "0.4", features = ["fuzz"] }
```

Example fuzz target:

```rust
#![no_main]
use libfuzzer_sys::fuzz_target;
use bigdecimal::BigDecimal;

fuzz_target!(|data: BigDecimal| {
    // Your fuzzing logic here
    let _ = data.to_string();
    let _ = data.sqrt();
});
```

Closes #122

### CODE REVIEW

**Comment by LucaCappelletti94** [comment]
For added context, this PR is needed to allow for adding `Arbitrary` derives in sqlparser, as shown here: https://github.com/apache/datafusion-sqlparser-rs/pull/2193

File: src/impl_arbitrary.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+impl<'a> Arbitrary<'a> for BigDecimal {
+    fn arbitrary(u: &mut Unstructured<'a>) -> ArbitraryResult<Self> {
+        let int_val = num_bigint::BigInt::arbitrary(u)?;
+        let scale = i64::arbitrary(u)? % SCALE_LIMIT;
+        Ok(BigDecimal::new(int_val, scale))
+    }
+
+    fn arbitrary_take_rest(mut u: Unstructured<'a>) -> ArbitraryResult<Self> {
+        // Use arbitrary for scale first (fixed size), then take rest for BigInt
+        let scale = i64::arbitrary(&mut u)? % SCALE_LIMIT;
```
Comment (akubera) [inline src/impl_arbitrary.rs:26]: Should the arbitrary scale be symmetric about zero, or should it be biased positive which indicates a positive number of decimal points which is the more likely scenario for using these? Or does it not matter for Arbitrary because every valid state should be representable?  And do we care that `0` is twice as likely as other numbers (1 out of a million vs 1 out of 2 million)?

I'm not familiar with Arbitrary so I might be overthinking this. I don't know what the fuzzers are looking for.

Comment (akubera) [reply src/impl_arbitrary.rs:26]: Also does order matter? `arbitrary` makes int first then scale, `arbitrary_take_rest` does scale then int. I feel like a well behaved impl would do those in the same order.

Comment (LucaCappelletti94) [reply src/impl_arbitrary.rs:26]: Regarding https://github.com/akubera/bigdecimal-rs/pull/163#discussion_r4057620084

I believe it should generate all possible valid states.

If implemented ideally, an `Arbitrary` impl of a downhill struct should take care of the distribution of the values depending of what values may be expected by the input stream of the downhill code.

Using the derive version should give a "reasonable baseline", and for instance it does not guarantee any self-consistent struct being generated by composition (which in certain harnesses it is the point as it forces the code to deal with any conceivable input it accepts). This would include a range-like struct that has start > end, even when all public methods to construct this struct have a Result that checks that start <= end. In such cases if the dev wants a self-consistent struct they need to write their own impl and avoid the derive.

Comment (LucaCappelletti94) [reply src/impl_arbitrary.rs:26]: Regarding https://github.com/akubera/bigdecimal-rs/pull/163#discussion_r4057634244

Consistency is better, I will adjust that and also fix a possible regression corner case that I noticed was uncovered in the test. It is not a problem at this time, but better safe than sorry.

**Review by akubera — APPROVED** [review]

**Comment by akubera** [comment]
Ok approved. Thanks a lot. Sorry I put this down for so many months.

**Comment by LucaCappelletti94** [comment]
> Ok approved. Thanks a lot. Sorry I put this down for so many months.

No worries, it'll be useful for fuzzing downhill!
